{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad820376-7fcd-4808-a3f2-7d2cb2663fcb",
   "metadata": {},
   "source": [
    "# **Bagian 1: Menginstal Paket yang Diperlukan** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "50f50732-c958-4973-b619-b2aa517d3479",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.11.9\n",
      "Requirement already satisfied: pip in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (24.0)\n",
      "Requirement already satisfied: setuptools in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (70.0.0)\n",
      "Requirement already satisfied: wheel in c:\\programdata\\anaconda3\\lib\\site-packages (0.43.0)\n",
      "Requirement already satisfied: cython in c:\\programdata\\anaconda3\\lib\\site-packages (3.0.10)\n",
      "Collecting PyYAML==6.0\n",
      "  Downloading PyYAML-6.0-cp311-cp311-win_amd64.whl.metadata (2.1 kB)\n",
      "Downloading PyYAML-6.0-cp311-cp311-win_amd64.whl (143 kB)\n",
      "   ---------------------------------------- 0.0/143.2 kB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/143.2 kB ? eta -:--:--\n",
      "   -- ------------------------------------- 10.2/143.2 kB ? eta -:--:--\n",
      "   -------- ------------------------------ 30.7/143.2 kB 435.7 kB/s eta 0:00:01\n",
      "   -------- ------------------------------ 30.7/143.2 kB 435.7 kB/s eta 0:00:01\n",
      "   ------------------- ------------------- 71.7/143.2 kB 393.8 kB/s eta 0:00:01\n",
      "   ------------------- ------------------- 71.7/143.2 kB 393.8 kB/s eta 0:00:01\n",
      "   -------------------------------- ----- 122.9/143.2 kB 423.5 kB/s eta 0:00:01\n",
      "   ----------------------------------- -- 133.1/143.2 kB 437.3 kB/s eta 0:00:01\n",
      "   ----------------------------------- -- 133.1/143.2 kB 437.3 kB/s eta 0:00:01\n",
      "   -------------------------------------- 143.2/143.2 kB 354.5 kB/s eta 0:00:00\n",
      "Installing collected packages: PyYAML\n",
      "  Attempting uninstall: PyYAML\n",
      "    Found existing installation: PyYAML 6.0.1\n",
      "    Uninstalling PyYAML-6.0.1:\n",
      "      Successfully uninstalled PyYAML-6.0.1\n",
      "Successfully installed PyYAML-6.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: Failed to remove contents in a temporary directory 'C:\\ProgramData\\anaconda3\\Lib\\site-packages\\~aml'.\n",
      "  You can safely remove it manually.\n",
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "conda-repo-cli 1.0.88 requires PyYAML>=6.0.1, but you have pyyaml 6.0 which is incompatible.\n",
      "label-studio 1.12.1 requires jsonschema==3.2.0, but you have jsonschema 4.22.0 which is incompatible.\n",
      "label-studio 1.12.1 requires wheel<=0.40.0,>=0.38.1, but you have wheel 0.43.0 which is incompatible.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting numpy==1.23.5\n",
      "  Using cached numpy-1.23.5-cp311-cp311-win_amd64.whl.metadata (2.3 kB)\n",
      "Requirement already satisfied: tensorflow==2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (2.12.0)\n",
      "Requirement already satisfied: tensorflow-intel==2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow==2.12.0) (2.12.0)\n",
      "Requirement already satisfied: absl-py>=1.0.0 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.1.0)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.6.3)\n",
      "Requirement already satisfied: flatbuffers>=2.0 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (24.3.25)\n",
      "Requirement already satisfied: gast<=0.4.0,>=0.2.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.4.0)\n",
      "Requirement already satisfied: google-pasta>=0.1.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.2.0)\n",
      "Requirement already satisfied: h5py>=2.9.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (3.11.0)\n",
      "Requirement already satisfied: jax>=0.3.15 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.4.28)\n",
      "Requirement already satisfied: libclang>=13.0.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (18.1.1)\n",
      "Requirement already satisfied: opt-einsum>=2.3.2 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (3.3.0)\n",
      "Requirement already satisfied: packaging in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (23.2)\n",
      "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (4.25.3)\n",
      "Requirement already satisfied: setuptools in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (70.0.0)\n",
      "Requirement already satisfied: six>=1.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.16.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.4.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.6 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (4.11.0)\n",
      "Requirement already satisfied: wrapt<1.15,>=1.11.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.14.1)\n",
      "Requirement already satisfied: grpcio<2.0,>=1.24.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.64.0)\n",
      "Requirement already satisfied: tensorboard<2.13,>=2.12 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.12.3)\n",
      "Requirement already satisfied: tensorflow-estimator<2.13,>=2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.12.0)\n",
      "Requirement already satisfied: keras<2.13,>=2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.12.0)\n",
      "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.31.0)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from astunparse>=1.6.0->tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.43.0)\n",
      "Requirement already satisfied: ml-dtypes>=0.2.0 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from jax>=0.3.15->tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.4.0)\n",
      "Requirement already satisfied: scipy>=1.9 in c:\\programdata\\anaconda3\\lib\\site-packages (from jax>=0.3.15->tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.13.0)\n",
      "Requirement already satisfied: google-auth<3,>=1.6.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.29.0)\n",
      "Requirement already satisfied: google-auth-oauthlib<1.1,>=0.5 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.0.0)\n",
      "Requirement already satisfied: markdown>=2.6.8 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (3.4.1)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.31.0)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.7.2)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.2.3)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (5.3.3)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in c:\\programdata\\anaconda3\\lib\\site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (4.9)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from google-auth-oauthlib<1.1,>=0.5->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.0.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (1.26.18)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2024.6.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from werkzeug>=1.0.1->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (2.1.3)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in c:\\programdata\\anaconda3\\lib\\site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1.1,>=0.5->tensorboard<2.13,>=2.12->tensorflow-intel==2.12.0->tensorflow==2.12.0) (3.2.2)\n",
      "Using cached numpy-1.23.5-cp311-cp311-win_amd64.whl (14.6 MB)\n",
      "Installing collected packages: numpy\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.26.4\n",
      "    Uninstalling numpy-1.26.4:\n",
      "      Successfully uninstalled numpy-1.26.4\n",
      "Successfully installed numpy-1.23.5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: Failed to remove contents in a temporary directory 'C:\\Users\\Muhammad Prakarsa\\AppData\\Local\\Temp\\pip-uninstall-0c80ch_0'.\n",
      "  You can safely remove it manually.\n",
      "  WARNING: Failed to remove contents in a temporary directory 'C:\\ProgramData\\anaconda3\\Lib\\site-packages\\~umpy.libs'.\n",
      "  You can safely remove it manually.\n",
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "gensim 4.3.0 requires FuzzyTM>=0.4.0, which is not installed.\n",
      "label-studio 1.12.1 requires jsonschema==3.2.0, but you have jsonschema 4.22.0 which is incompatible.\n",
      "label-studio 1.12.1 requires numpy<2.0.0,>=1.24.3, but you have numpy 1.23.5 which is incompatible.\n",
      "label-studio 1.12.1 requires wheel<=0.40.0,>=0.38.1, but you have wheel 0.43.0 which is incompatible.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting jsonschema==3.2.0\n",
      "  Using cached jsonschema-3.2.0-py2.py3-none-any.whl.metadata (7.8 kB)\n",
      "Collecting wheel==0.40.0\n",
      "  Using cached wheel-0.40.0-py3-none-any.whl.metadata (2.1 kB)\n",
      "Requirement already satisfied: attrs>=17.4.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from jsonschema==3.2.0) (23.1.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from jsonschema==3.2.0) (0.20.0)\n",
      "Requirement already satisfied: setuptools in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from jsonschema==3.2.0) (70.0.0)\n",
      "Requirement already satisfied: six>=1.11.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from jsonschema==3.2.0) (1.16.0)\n",
      "Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)\n",
      "Using cached wheel-0.40.0-py3-none-any.whl (64 kB)\n",
      "Installing collected packages: wheel, jsonschema\n",
      "  Attempting uninstall: wheel\n",
      "    Found existing installation: wheel 0.43.0\n",
      "    Uninstalling wheel-0.43.0:\n",
      "      Successfully uninstalled wheel-0.43.0\n",
      "  Attempting uninstall: jsonschema\n",
      "    Found existing installation: jsonschema 4.22.0\n",
      "    Uninstalling jsonschema-4.22.0:\n",
      "      Successfully uninstalled jsonschema-4.22.0\n",
      "Successfully installed jsonschema-3.2.0 wheel-0.40.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "conda-build 24.5.0 requires jsonschema>=4.19, but you have jsonschema 3.2.0 which is incompatible.\n",
      "jupyterlab-server 2.25.1 requires jsonschema>=4.18.0, but you have jsonschema 3.2.0 which is incompatible.\n",
      "jupyter-events 0.8.0 requires jsonschema[format-nongpl]>=4.18.0, but you have jsonschema 3.2.0 which is incompatible.\n",
      "label-studio 1.12.1 requires numpy<2.0.0,>=1.24.3, but you have numpy 1.23.5 which is incompatible.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tensorflow-addons\n",
      "  Using cached tensorflow_addons-0.22.0-cp311-cp311-win_amd64.whl.metadata (1.8 kB)\n",
      "Collecting tensorflow-datasets\n",
      "  Using cached tensorflow_datasets-4.9.6-py3-none-any.whl.metadata (9.5 kB)\n",
      "Collecting tensorflow-hub\n",
      "  Using cached tensorflow_hub-0.16.1-py2.py3-none-any.whl.metadata (1.3 kB)\n",
      "Collecting tensorflow-model-optimization\n",
      "  Using cached tensorflow_model_optimization-0.8.0-py2.py3-none-any.whl.metadata (904 bytes)\n",
      "Collecting tf-models-official==2.13.1\n",
      "  Using cached tf_models_official-2.13.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: Cython in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (3.0.10)\n",
      "Requirement already satisfied: Pillow in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (10.3.0)\n",
      "Collecting gin-config (from tf-models-official==2.13.1)\n",
      "  Using cached gin_config-0.5.0-py3-none-any.whl.metadata (2.9 kB)\n",
      "Collecting google-api-python-client>=1.6.7 (from tf-models-official==2.13.1)\n",
      "  Using cached google_api_python_client-2.133.0-py2.py3-none-any.whl.metadata (6.7 kB)\n",
      "Collecting immutabledict (from tf-models-official==2.13.1)\n",
      "  Using cached immutabledict-4.2.0-py3-none-any.whl.metadata (3.4 kB)\n",
      "Collecting kaggle>=1.3.9 (from tf-models-official==2.13.1)\n",
      "  Using cached kaggle-1.6.14.tar.gz (82 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: matplotlib in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (3.8.4)\n",
      "Requirement already satisfied: numpy>=1.20 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (1.23.5)\n",
      "Collecting oauth2client (from tf-models-official==2.13.1)\n",
      "  Using cached oauth2client-4.1.3-py2.py3-none-any.whl.metadata (1.2 kB)\n",
      "Collecting opencv-python-headless (from tf-models-official==2.13.1)\n",
      "  Using cached opencv_python_headless-4.10.0.82-cp37-abi3-win_amd64.whl.metadata (20 kB)\n",
      "Requirement already satisfied: pandas>=0.22.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (2.2.1)\n",
      "Requirement already satisfied: psutil>=5.4.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (5.9.0)\n",
      "Requirement already satisfied: py-cpuinfo>=3.3.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official==2.13.1) (9.0.0)\n",
      "Collecting pycocotools (from tf-models-official==2.13.1)\n",
      "  Using cached pycocotools-2.0.7-cp311-cp311-win_amd64.whl.metadata (1.1 kB)\n",
      "Collecting pyyaml<5.4.0,>=5.1 (from tf-models-official==2.13.1)\n",
      "  Using cached PyYAML-5.3.1.tar.gz (269 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'error'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  python setup.py egg_info did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [37 lines of output]\n",
      "  Traceback (most recent call last):\n",
      "    File \"<string>\", line 2, in <module>\n",
      "    File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Local\\Temp\\pip-install-gb0rjs1u\\pyyaml_66ae6f8c5dbc44089e810a57d380d2da\\setup.py\", line 291, in <module>\n",
      "      setup(\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\core.py\", line 184, in setup\n",
      "      return run_commands(dist)\n",
      "             ^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\core.py\", line 200, in run_commands\n",
      "      dist.run_commands()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\dist.py\", line 969, in run_commands\n",
      "      self.run_command(cmd)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\dist.py\", line 968, in run_command\n",
      "      super().run_command(command)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "      cmd_obj.run()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 321, in run\n",
      "      self.find_sources()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 329, in find_sources\n",
      "      mm.run()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 550, in run\n",
      "      self.add_defaults()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 588, in add_defaults\n",
      "      sdist.add_defaults(self)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\sdist.py\", line 102, in add_defaults\n",
      "      super().add_defaults()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\command\\sdist.py\", line 250, in add_defaults\n",
      "      self._add_defaults_ext()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\command\\sdist.py\", line 335, in _add_defaults_ext\n",
      "      self.filelist.extend(build_ext.get_source_files())\n",
      "                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Local\\Temp\\pip-install-gb0rjs1u\\pyyaml_66ae6f8c5dbc44089e810a57d380d2da\\setup.py\", line 199, in get_source_files\n",
      "      self.cython_sources(ext.sources, ext)\n",
      "      ^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\cmd.py\", line 107, in __getattr__\n",
      "      raise AttributeError(attr)\n",
      "  AttributeError: cython_sources\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "error: metadata-generation-failed\n",
      "\n",
      "Encountered error while generating package metadata.\n",
      "\n",
      "See above for output.\n",
      "\n",
      "note: This is an issue with the package mentioned above, not pip.\n",
      "hint: See above for details.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting mediapipe-model-maker\n",
      "  Using cached mediapipe_model_maker-0.2.1.4-py3-none-any.whl.metadata (1.7 kB)\n",
      "Requirement already satisfied: absl-py in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe-model-maker) (2.1.0)\n",
      "Requirement already satisfied: mediapipe>=0.10.0 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe-model-maker) (0.10.14)\n",
      "Requirement already satisfied: numpy in c:\\programdata\\anaconda3\\lib\\site-packages (from mediapipe-model-maker) (1.23.5)\n",
      "Requirement already satisfied: opencv-python in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe-model-maker) (4.9.0.80)\n",
      "Requirement already satisfied: tensorflow<2.16,>=2.10 in c:\\programdata\\anaconda3\\lib\\site-packages (from mediapipe-model-maker) (2.12.0)\n",
      "Collecting tensorflow-addons (from mediapipe-model-maker)\n",
      "  Using cached tensorflow_addons-0.22.0-cp311-cp311-win_amd64.whl.metadata (1.8 kB)\n",
      "Collecting tensorflow-datasets (from mediapipe-model-maker)\n",
      "  Using cached tensorflow_datasets-4.9.6-py3-none-any.whl.metadata (9.5 kB)\n",
      "Collecting tensorflow-hub (from mediapipe-model-maker)\n",
      "  Using cached tensorflow_hub-0.16.1-py2.py3-none-any.whl.metadata (1.3 kB)\n",
      "Collecting tensorflow-model-optimization<0.8.0 (from mediapipe-model-maker)\n",
      "  Using cached tensorflow_model_optimization-0.7.5-py2.py3-none-any.whl.metadata (914 bytes)\n",
      "INFO: pip is looking at multiple versions of mediapipe-model-maker to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting mediapipe-model-maker\n",
      "  Using cached mediapipe_model_maker-0.2.1.3-py3-none-any.whl.metadata (1.6 kB)\n",
      "Collecting tf-models-official>=2.13.1 (from mediapipe-model-maker)\n",
      "  Using cached tf_models_official-2.16.0-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: attrs>=19.1.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (23.1.0)\n",
      "Requirement already satisfied: flatbuffers>=2.0 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (24.3.25)\n",
      "Requirement already satisfied: jax in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (0.4.28)\n",
      "Requirement already satisfied: jaxlib in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (0.4.28)\n",
      "Requirement already satisfied: matplotlib in c:\\programdata\\anaconda3\\lib\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (3.8.4)\n",
      "Requirement already satisfied: opencv-contrib-python in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (4.9.0.80)\n",
      "Requirement already satisfied: protobuf<5,>=4.25.3 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (4.25.3)\n",
      "Requirement already satisfied: sounddevice>=0.4.4 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from mediapipe>=0.10.0->mediapipe-model-maker) (0.4.7)\n",
      "Requirement already satisfied: tensorflow-intel==2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow<2.16,>=2.10->mediapipe-model-maker) (2.12.0)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (1.6.3)\n",
      "Requirement already satisfied: gast<=0.4.0,>=0.2.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (0.4.0)\n",
      "Requirement already satisfied: google-pasta>=0.1.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (0.2.0)\n",
      "Requirement already satisfied: h5py>=2.9.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (3.11.0)\n",
      "Requirement already satisfied: libclang>=13.0.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (18.1.1)\n",
      "Requirement already satisfied: opt-einsum>=2.3.2 in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (3.3.0)\n",
      "Requirement already satisfied: packaging in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (23.2)\n",
      "Requirement already satisfied: setuptools in c:\\users\\muhammad prakarsa\\appdata\\roaming\\python\\python311\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (70.0.0)\n",
      "Requirement already satisfied: six>=1.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (1.16.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (2.4.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.6 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (4.11.0)\n",
      "Requirement already satisfied: wrapt<1.15,>=1.11.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (1.14.1)\n",
      "Requirement already satisfied: grpcio<2.0,>=1.24.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (1.64.0)\n",
      "Requirement already satisfied: tensorboard<2.13,>=2.12 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (2.12.3)\n",
      "Requirement already satisfied: tensorflow-estimator<2.13,>=2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (2.12.0)\n",
      "Requirement already satisfied: keras<2.13,>=2.12.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (2.12.0)\n",
      "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tensorflow-intel==2.12.0->tensorflow<2.16,>=2.10->mediapipe-model-maker) (0.31.0)\n",
      "Requirement already satisfied: Cython in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (3.0.10)\n",
      "Requirement already satisfied: Pillow in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (10.3.0)\n",
      "Collecting gin-config (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached gin_config-0.5.0-py3-none-any.whl.metadata (2.9 kB)\n",
      "Collecting google-api-python-client>=1.6.7 (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached google_api_python_client-2.133.0-py2.py3-none-any.whl.metadata (6.7 kB)\n",
      "Collecting immutabledict (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached immutabledict-4.2.0-py3-none-any.whl.metadata (3.4 kB)\n",
      "Collecting kaggle>=1.3.9 (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached kaggle-1.6.14.tar.gz (82 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Collecting oauth2client (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached oauth2client-4.1.3-py2.py3-none-any.whl.metadata (1.2 kB)\n",
      "Collecting opencv-python-headless (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached opencv_python_headless-4.10.0.82-cp37-abi3-win_amd64.whl.metadata (20 kB)\n",
      "Requirement already satisfied: pandas>=0.22.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (2.2.1)\n",
      "Requirement already satisfied: psutil>=5.4.3 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (5.9.0)\n",
      "Requirement already satisfied: py-cpuinfo>=3.3.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (9.0.0)\n",
      "Collecting pycocotools (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached pycocotools-2.0.7-cp311-cp311-win_amd64.whl.metadata (1.1 kB)\n",
      "Requirement already satisfied: pyyaml>=6.0.0 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (6.0)\n",
      "Collecting sacrebleu (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached sacrebleu-2.4.2-py3-none-any.whl.metadata (58 kB)\n",
      "Requirement already satisfied: scipy>=0.19.1 in c:\\programdata\\anaconda3\\lib\\site-packages (from tf-models-official>=2.13.1->mediapipe-model-maker) (1.13.0)\n",
      "Collecting sentencepiece (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached sentencepiece-0.2.0-cp311-cp311-win_amd64.whl.metadata (8.3 kB)\n",
      "Collecting seqeval (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached seqeval-1.2.2.tar.gz (43 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Collecting tensorflow-model-optimization>=0.4.1 (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached tensorflow_model_optimization-0.8.0-py2.py3-none-any.whl.metadata (904 bytes)\n",
      "INFO: pip is looking at multiple versions of tf-models-official to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting tf-models-official>=2.13.1 (from mediapipe-model-maker)\n",
      "  Using cached tf_models_official-2.15.0-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "  Using cached tf_models_official-2.14.2-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "  Using cached tf_models_official-2.14.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "  Using cached tf_models_official-2.14.0-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "  Using cached tf_models_official-2.13.2-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "  Using cached tf_models_official-2.13.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Collecting pyyaml<5.4.0,>=5.1 (from tf-models-official>=2.13.1->mediapipe-model-maker)\n",
      "  Using cached PyYAML-5.3.1.tar.gz (269 kB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'error'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  python setup.py egg_info did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [37 lines of output]\n",
      "  Traceback (most recent call last):\n",
      "    File \"<string>\", line 2, in <module>\n",
      "    File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Local\\Temp\\pip-install-rrx6p02q\\pyyaml_dca51ab683794017a52f224aecedf41a\\setup.py\", line 291, in <module>\n",
      "      setup(\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\core.py\", line 184, in setup\n",
      "      return run_commands(dist)\n",
      "             ^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\core.py\", line 200, in run_commands\n",
      "      dist.run_commands()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\dist.py\", line 969, in run_commands\n",
      "      self.run_command(cmd)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\dist.py\", line 968, in run_command\n",
      "      super().run_command(command)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "      cmd_obj.run()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 321, in run\n",
      "      self.find_sources()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 329, in find_sources\n",
      "      mm.run()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 550, in run\n",
      "      self.add_defaults()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\egg_info.py\", line 588, in add_defaults\n",
      "      sdist.add_defaults(self)\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\command\\sdist.py\", line 102, in add_defaults\n",
      "      super().add_defaults()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\command\\sdist.py\", line 250, in add_defaults\n",
      "      self._add_defaults_ext()\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\command\\sdist.py\", line 335, in _add_defaults_ext\n",
      "      self.filelist.extend(build_ext.get_source_files())\n",
      "                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Local\\Temp\\pip-install-rrx6p02q\\pyyaml_dca51ab683794017a52f224aecedf41a\\setup.py\", line 199, in get_source_files\n",
      "      self.cython_sources(ext.sources, ext)\n",
      "      ^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Muhammad Prakarsa\\AppData\\Roaming\\Python\\Python311\\site-packages\\setuptools\\_distutils\\cmd.py\", line 107, in __getattr__\n",
      "      raise AttributeError(attr)\n",
      "  AttributeError: cython_sources\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "error: metadata-generation-failed\n",
      "\n",
      "Encountered error while generating package metadata.\n",
      "\n",
      "See above for output.\n",
      "\n",
      "note: This is an issue with the package mentioned above, not pip.\n",
      "hint: See above for details.\n"
     ]
    }
   ],
   "source": [
    "# Menggunakan magics Jupyter untuk menjalankan perintah shell\n",
    "!python --version\n",
    "!pip install --upgrade pip setuptools wheel\n",
    "!pip install cython\n",
    "!pip install PyYAML==6.0\n",
    "!pip install numpy==1.23.5 tensorflow==2.12.0\n",
    "!pip install jsonschema==3.2.0 wheel==0.40.0\n",
    "!pip install tensorflow-addons tensorflow-datasets tensorflow-hub tensorflow-model-optimization tf-models-official==2.13.1\n",
    "!pip install mediapipe-model-maker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1fc7614-3591-4c65-b014-d81ba54363e1",
   "metadata": {},
   "source": [
    "# **Bagian 2: Mengimpor Library yang Diperlukan** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f88adf55-0c59-4d0b-b308-e01ed3748d68",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'mediapipe_model_maker'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[5], line 7\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtensorflow\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mtf\u001b[39;00m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m tf\u001b[38;5;241m.\u001b[39m__version__\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m2\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m----> 7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mmediapipe_model_maker\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m object_detector\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'mediapipe_model_maker'"
     ]
    }
   ],
   "source": [
    "# Import library yang dibutuhkan\n",
    "import os\n",
    "import json\n",
    "import tensorflow as tf\n",
    "assert tf.__version__.startswith('2')\n",
    "\n",
    "from mediapipe_model_maker import object_detector"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "668db956-c86e-47fd-929e-3ce24ad8febc",
   "metadata": {},
   "source": [
    "# **Bagian 3: Menyiapkan Dataset**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cc04e8a-f4a2-4ff5-89b8-0e3201d71e9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Definisikan jalur dataset\n",
    "train_dataset_path = \"Dataset500/train\"\n",
    "validation_dataset_path = \"Dataset500/validation\"\n",
    "\n",
    "# Membaca dan menampilkan label dari dataset pelatihan\n",
    "with open(os.path.join(train_dataset_path, \"result.json\"), \"r\") as f:\n",
    "    labels_json = json.load(f)\n",
    "\n",
    "for category_item in labels_json[\"categories\"]:\n",
    "    print(f\"{category_item['id']}: {category_item['name']}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c360bf87-0143-45f5-b868-e3716119799b",
   "metadata": {},
   "source": [
    "# **Bagian 4: Fungsi untuk Visualisasi Dataset**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca58801-a0d2-40de-b689-02506f29f922",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fungsi untuk visualisasi dataset pelatihan\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import patches, text, patheffects\n",
    "from collections import defaultdict\n",
    "import math\n",
    "\n",
    "def draw_outline(obj):\n",
    "    obj.set_path_effects([patheffects.Stroke(linewidth=4,  foreground='black'), patheffects.Normal()])\n",
    "\n",
    "def draw_box(ax, bb):\n",
    "    patch = ax.add_patch(patches.Rectangle((bb[0], bb[1]), bb[2], bb[3], fill=False, edgecolor='red', lw=2))\n",
    "    draw_outline(patch)\n",
    "\n",
    "def draw_text(ax, bb, txt, disp):\n",
    "    text = ax.text(bb[0], (bb[1]-disp), txt, verticalalignment='top', color='white', fontsize=10, weight='bold')\n",
    "    draw_outline(text)\n",
    "\n",
    "def draw_bbox(ax, annotations_list, id_to_label, image_shape):\n",
    "    for annotation in annotations_list:\n",
    "        cat_id = annotation[\"category_id\"]\n",
    "        bbox = annotation[\"bbox\"]\n",
    "        draw_box(ax, bbox)\n",
    "        draw_text(ax, bbox, id_to_label[cat_id], image_shape[0] * 0.05)\n",
    "\n",
    "def visualize(dataset_folder, max_examples=None):\n",
    "    with open(os.path.join(dataset_folder, \"labels.json\"), \"r\") as f:\n",
    "        labels_json = json.load(f)\n",
    "\n",
    "    images = labels_json[\"images\"]\n",
    "    cat_id_to_label = {item[\"id\"]: item[\"name\"] for item in labels_json[\"categories\"]}\n",
    "    image_annots = defaultdict(list)\n",
    "    for annotation_obj in labels_json[\"annotations\"]:\n",
    "        image_id = annotation_obj[\"image_id\"]\n",
    "        image_annots[image_id].append(annotation_obj)\n",
    "\n",
    "    if max_examples is None:\n",
    "        max_examples = len(image_annots.items())\n",
    "\n",
    "    n_rows = math.ceil(max_examples / 3)\n",
    "    fig, axs = plt.subplots(n_rows, 3, figsize=(24, n_rows*8))\n",
    "\n",
    "    for ind, (image_id, annotations_list) in enumerate(list(image_annots.items())[:max_examples]):\n",
    "        ax = axs[ind//3, ind%3]\n",
    "        img = plt.imread(os.path.join(dataset_folder, \"images\", images[image_id][\"file_name\"]))\n",
    "        ax.imshow(img)\n",
    "        draw_bbox(ax, annotations_list, cat_id_to_label, img.shape)\n",
    "\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36ef324-54e6-4cc5-aa21-662809e5edf5",
   "metadata": {},
   "source": [
    "# **Bagian 5: Visualisasi Dataset**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ccbbe35-f566-4d17-bd5c-262b2882c1e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "visualize(train_dataset_path, 5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b55c195-f497-4c6f-978c-0e50b3687662",
   "metadata": {},
   "source": [
    "# **Bagian 6: Memuat Data Pelatihan dan Validasi**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae167bb-df82-4a54-bcda-5793be342499",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Memuat data pelatihan dan validasi\n",
    "train_data = object_detector.Dataset.from_coco_folder(train_dataset_path, cache_dir=\"/tmp/od_data/train\")\n",
    "validation_data = object_detector.Dataset.from_coco_folder(validation_dataset_path, cache_dir=\"/tmp/od_data/validation\")\n",
    "\n",
    "print(\"train_data size: \", train_data.size)\n",
    "print(\"validation_data size: \", validation_data.size)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c35bef-fc19-4d67-ab5f-96081faa0d71",
   "metadata": {},
   "source": [
    "# **Bagian 7: Mengonfigurasi dan Melatih Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23f0025c-eec1-465b-afa6-558343ff879b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mengonfigurasi model\n",
    "spec = object_detector.SupportedModels.MOBILENET_MULTI_AVG\n",
    "hparams = object_detector.HParams(export_dir='exported_model')\n",
    "options = object_detector.ObjectDetectorOptions(\n",
    "    supported_model=spec,\n",
    "    hparams=hparams\n",
    ")\n",
    "\n",
    "# Melatih model\n",
    "model = object_detector.ObjectDetector.create(\n",
    "    train_data=train_data,\n",
    "    validation_data=validation_data,\n",
    "    options=options\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06d45288-b32e-400a-adc7-5831479bab99",
   "metadata": {},
   "source": [
    "# **Bagian 8: Mengevaluasi Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "631980b1-762c-48b3-aad7-d5641496ef86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mengevaluasi model\n",
    "loss, coco_metrics = model.evaluate(validation_data, batch_size=4)\n",
    "print(f\"Validation loss: {loss}\")\n",
    "print(f\"Validation coco metrics: {coco_metrics}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5056c3e6-4618-4bb4-a9f1-caf3ffa0379d",
   "metadata": {},
   "source": [
    "# **Bagian 9: Mengekspor dan Mengunduh Model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b64b98-4fc9-4722-96ed-dae4e30a2ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mengekspor model\n",
    "model.export_model()\n",
    "\n",
    "# Menampilkan hasil ekspor\n",
    "!ls exported_model\n",
    "\n",
    "# Mengunduh model yang telah diekspor\n",
    "from google.colab import files\n",
    "files.download('exported_model/model.tflite')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
